Include Host header for CONNECT proxy requests

soedar opened the pull request:
We have encountered an issue where we would get HTTP 400 when we specify a proxy when making a HTTP request:
```
julia> import HTTP

julia> HTTP.request("GET","https://example.com", proxy="http://proxy:80")
ERROR: HTTP.ExceptionRequest.StatusError(400, "CONNECT", "example.com:443", HTTP.Messages.Response:
"""
HTTP/1.1 400 Bad Request
Date: Wed, 19 May 2021 16:15:03 GMT
Server: Apache
Strict-Transport-Security: max-age=63072000; includeSubdomains; preload
X-Frame-Options: SAMEORIGIN
Content-Length: 226
Connection: close
Content-Type: text/html; charset=iso-8859-1

""")
Stacktrace:
 [1] request(::Type{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer{Union{}}}}}, ::URIs.URI, ::Vararg{Any, N} where N; kw::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:iofunction, :reached_redirect_limit, :proxy), Tuple{Nothing, Bool, String}}})
   @ HTTP.ExceptionRequest ~/.julia/packages/HTTP/cxgat/src/ExceptionRequest.jl:22
 [2] (::Base.var"#70#72"{Base.var"#70#71#73"{ExponentialBackOff, HTTP.RetryRequest.var"#2#3"{Bool, HTTP.Messages.Request}, typeof(HTTP.request)}})(::Type, ::Vararg{Any, N} where N; kwargs::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:iofunction, :reached_redirect_limit, :proxy), Tuple{Nothing, Bool, String}}})
   @ Base ./error.jl:288
 [3] #request#1
   @ ~/.julia/packages/HTTP/cxgat/src/RetryRequest.jl:44 [inlined]
 [4] request(::Type{HTTP.MessageRequest.MessageLayer{HTTP.RetryRequest.RetryLayer{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer{Union{}}}}}}}, method::String, url::URIs.URI, headers::Vector{Pair{SubString{String}, SubString{String}}}, body::Vector{UInt8}; http_version::VersionNumber, target::String, parent::Nothing, iofunction::Nothing, kw::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:reached_redirect_limit, :proxy), Tuple{Bool, String}}})
   @ HTTP.MessageRequest ~/.julia/packages/HTTP/cxgat/src/MessageRequest.jl:66
 [5] request(::Type{HTTP.BasicAuthRequest.BasicAuthLayer{HTTP.MessageRequest.MessageLayer{HTTP.RetryRequest.RetryLayer{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer{Union{}}}}}}}}, method::String, url::URIs.URI, headers::Vector{Pair{SubString{String}, SubString{String}}}, body::Vector{UInt8}; kw::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:reached_redirect_limit, :proxy), Tuple{Bool, String}}})
   @ HTTP.BasicAuthRequest ~/.julia/packages/HTTP/cxgat/src/BasicAuthRequest.jl:28
 [6] request(::Type{HTTP.RedirectRequest.RedirectLayer{HTTP.BasicAuthRequest.BasicAuthLayer{HTTP.MessageRequest.MessageLayer{HTTP.RetryRequest.RetryLayer{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer{Union{}}}}}}}}}, method::String, url::URIs.URI, headers::Vector{Pair{SubString{String}, SubString{String}}}, body::Vector{UInt8}; redirect_limit::Int64, forwardheaders::Bool, kw::Base.Iterators.Pairs{Symbol, String, Tuple{Symbol}, NamedTuple{(:proxy,), Tuple{String}}})
   @ HTTP.RedirectRequest ~/.julia/packages/HTTP/cxgat/src/RedirectRequest.jl:24
 [7] request(method::String, url::String, h::Vector{Pair{SubString{String}, SubString{String}}}, b::Vector{UInt8}; headers::Vector{Pair{SubString{String}, SubString{String}}}, body::Vector{UInt8}, query::Nothing, kw::Base.Iterators.Pairs{Symbol, String, Tuple{Symbol}, NamedTuple{(:proxy,), Tuple{String}}})
   @ HTTP ~/.julia/packages/HTTP/cxgat/src/HTTP.jl:315
 [8] top-level scope
   @ REPL[2]:1
```

Turns out that our http proxy server (apache configured as a forward proxy) would return HTTP 400 when the Host header is not specified as part of CONNECT request. The following is observed in the apache logs:

```
httpd_1            | [Wed May 19 16:15:03.242807 2021] [core:debug] [pid 15:tid 140522410957600] protocol.c(1445): [client <ip>:53238] AH00569: client sent HTTP/1.1 request without hostname (see RFC2616 section 14.23): /
httpd_1            | [Wed May 19 16:15:03.242840 2021] [headers:debug] [pid 15:tid 140522410957600] mod_headers.c(899): AH01503: headers: ap_headers_error_filter()
httpd_1            | <ip> - - [19/May/2021:16:15:03 +0000] "CONNECT example.com:443 HTTP/1.1" 400 0 "-" "-" 14375 0
```

From the [RFC2616 section 14.23](https://datatracker.ietf.org/doc/html/rfc2616#section-14.23):
```
   A client MUST include a Host header field in all HTTP/1.1 request
   messages . If the requested URI does not include an Internet host
   name for the service being requested, then the Host header field MUST
   be given with an empty value. An HTTP/1.1 proxy MUST ensure that any
   request message it forwards does contain an appropriate Host header
   field that identifies the service being requested by the proxy.
```

---

This PR fixes the issue by including the original `Host` header in the request. After this change, Apache correctly accepts the CONNECT requests and forwards the HTTP request

```
julia> HTTP.request("GET","https://example.com", proxy="http://proxy:80")
HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
Accept-Ranges: bytes
Age: 343609
Cache-Control: max-age=604800
Content-Type: text/html; charset=UTF-8
Date: Wed, 19 May 2021 16:17:05 GMT
Etag: "3147526947"
Expires: Wed, 26 May 2021 16:17:05 GMT
Last-Modified: Thu, 17 Oct 2019 07:18:26 GMT
Server: ECS (oxr/8324)
Vary: Accept-Encoding
X-Cache: HIT
Content-Length: 1256
<body truncated>
"""

# Apache logs:
httpd_1            | <ip> - - [19/May/2021:16:17:04 +0000] "CONNECT example.com:443 HTTP/1.1" 200 6382 "-" "-" 1719673 0
```

fredrikekre commented:
Can you add a test? And perhaps also a note in the CHANGELOG.md file.

Teo-ShaoWei commented on test/client.jl:
```diff
@@ -407,3 +407,35 @@ end
+            sock = accept(proxy)
+            while isopen(sock)
+                line = readline(sock)
+                if line == ""
+                    break
+                end
```
We can consider:
```suggestion
                isempty(line) && break
```

soedar commented:
@fredrikekre I wasn't sure how best to test this, but I've added a simple test that checks if the Host header is being forwarded in when using http proxy. I've also updated the changelog with the fix under unreleased.

fredrikekre commented:
Should the port be included in the Host header too for `CONNECT` requests? Here is the netcat output of
```
$ export https_proxy=http://localhost:8080^C
$ curl https://julialang.org
```
which gives
```
$ netcat -lp 8080
CONNECT julialang.org:443 HTTP/1.1
Host: julialang.org:443
User-Agent: curl/7.68.0
Proxy-Connection: Keep-Alive
```
for example.

fredrikekre commented on src/ConnectionRequest.jl:
```diff
@@ -126,7 +126,7 @@ sockettype(url::URI, default) = url.scheme in ("wss", "https") ? SSLContext : de
     target = "$(URIs.hoststring(target_url.host)):$(target_url.port)"
     @debug 1 "📡  CONNECT HTTPS tunnel to $target"
-    headers = Dict(filter(x->x.first == "Proxy-Authorization", req.headers))
+    headers = Dict(filter(x->x.first in ("Host", "Proxy-Authorization"), req.headers))
```
Perhaps something like this? `target` is the host and port (from 2 lines above)
```suggestion
headers = Dict("Host" => target)
if (auth = header(req, "Proxy-Authorization"); !isempty(auth))
    headers["Proxy-Authorization"] = auth
end
```

soedar replied:
Thanks!

fredrikekre commented on CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
+### Fixed
+- Include Host header for CONNECT proxy requests ([#714])
```
```suggestion
- Include `Host` header for `CONNECT` proxy requests ([#714])
```

fredrikekre commented:
FYI, this is part of HTTP version 0.9.9 that was just released.

soedar commented:
Thanks for the fast review and release!
